wolfcrypt/test/test.c

```diff
@@ -3003,8 +3035,311 @@ WOLFSSL_TEST_SUBROUTINE int sha512_test(void)
+    c.inLen  = XSTRLEN(c.input);
+    c.outLen = WC_SHA512_224_DIGEST_SIZE;
+
+    test_sha[0] = a;
+    test_sha[1] = b;
+    test_sha[2] = c;
+
+    ret = wc_InitSha512_224_ex(&sha, HEAP_HINT, devId);
+    if (ret != 0)
+        return -22400;
+    ret = wc_InitSha512_224_ex(&shaCopy, HEAP_HINT, devId);
+    if (ret != 0) {
+        wc_Sha512_224Free(&sha);
+        return -22401;
+    }
+
+    for (i = 0; i < times; ++i) {
+        ret = wc_Sha512_224Update(&sha, (byte*)test_sha[i].input,
+            (word32)test_sha[i].inLen);
+        if (ret != 0)
+            ERROR_OUT(-22402 - i, exit);
+        ret = wc_Sha512_224GetHash(&sha, hashcopy);
+        if (ret != 0)
+            ERROR_OUT(-22403 - i, exit);
+        ret = wc_Sha512_224Copy(&sha, &shaCopy);
+        if (ret != 0)
+            ERROR_OUT(-22404 - i, exit);
+        ret = wc_Sha512_224Final(&sha, hash);
+        if (ret != 0)
+            ERROR_OUT(-22405 - i, exit);
+        wc_Sha512_224Free(&shaCopy);
+
+        if (XMEMCMP(hash, test_sha[i].output, WC_SHA512_224_DIGEST_SIZE) != 0)
+            ERROR_OUT(-22406 - i, exit);
+
+
+        if (XMEMCMP(hash, hashcopy, WC_SHA512_224_DIGEST_SIZE) != 0)
+            ERROR_OUT(-22407 - i, exit);
+    }
+
+#ifndef NO_LARGE_HASH_TEST
+    /* BEGIN LARGE HASH TEST */ {
+    byte large_input[1024];
+#ifdef HASH_SIZE_LIMIT
+    const char* large_digest =
+        "\x98\x68\xc3\xd9\xb9\xef\x17\x53"
+        "\x43\x66\x0e\x60\xdf\x29\xf8\xef"
+        "\x96\xe3\x93\x34\x8c\x6f\xc0\xeb"
+        "\x14\x6c\xcf\x6a";
+#else
+    const char* large_digest =
+        "\x26\x5f\x98\xd1\x76\x49\x71\x4e"
+        "\x82\xb7\x9d\x52\x32\x67\x9d\x56"
+        "\x91\xf5\x88\xc3\x05\xbb\x3f\x90"
+        "\xe2\x4e\x85\x05";
 #endif
 
+    for (i = 0; i < (int)sizeof(large_input); i++) {
+        large_input[i] = (byte)(i & 0xFF);
+    }
+#ifdef HASH_SIZE_LIMIT
+    times = 20;
+#else
+    times = 100;
+#endif
+    for (i = 0; i < times; ++i) {
+        ret = wc_Sha512_224Update(&sha, (byte*)large_input,
+            (word32)sizeof(large_input));
+        if (ret != 0)
+            ERROR_OUT(-22408, exit);
+    }
+    ret = wc_Sha512_224Final(&sha, hash);
+    if (ret != 0)
+        ERROR_OUT(-22409, exit);
+    if (XMEMCMP(hash, large_digest, WC_SHA512_224_DIGEST_SIZE) != 0)
+        ERROR_OUT(-22410, exit);
+
+#ifndef NO_UNALIGNED_MEMORY_TEST
+    /* Unaligned memory access test */
+    for (i = 1; i < 16; i++) {
+        ret = wc_Sha512_224Update(&sha, (byte*)large_input + i,
+            (word32)sizeof(large_input) - i);
+        if (ret != 0)
+            ERROR_OUT(-22411, exit);
+        ret = wc_Sha512_224Final(&sha, hash);
+    }
+#endif
+    } /* END LARGE HASH TEST */
+#endif /* NO_LARGE_HASH_TEST */
+
+exit:
+    wc_Sha512_224Free(&sha);
+    wc_Sha512_224Free(&shaCopy);
+
+    return ret;
+} /* sha512_224_test */
+#endif /* !defined(WOLFSSL_NOSHA512_224) ... */
+
+
+#if !defined(WOLFSSL_NOSHA512_256) && \
+    !defined(HAVE_FIPS)
```

Comment:
Same. Why not `(!defined(HAVE_FIPS) || FIPS_VERSION_GE(5, 2)) && !defined(HAVE_SELFTEST)`?